### PR TITLE
Rewrite Enum.CompareTo in C#

### PR DESCRIPTION
**Pull request by jkotas**

### CODE REVIEW

File: src/System.Private.CoreLib/shared/System/Enum.cs [hunk 1]
```diff
@@ -906,6 +906,57 @@ public override string ToString()
...
+            if (this.GetType() != target.GetType())
+                throw new ArgumentException(SR.Format(SR.Arg_EnumAndObjectMustBeSameType, target.GetType(), this.GetType()));
+
+            ref byte pThisValue = ref this.GetRawData();
+            ref byte pTargetValue = ref target.GetRawData();
+
+            switch (this.InternalGetCorElementType())
+            {
+                case CorElementType.ELEMENT_TYPE_I1:
+                    return Unsafe.As<byte, sbyte>(ref pThisValue).CompareTo(Unsafe.As<byte, sbyte>(ref pTargetValue));
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Enum.cs]: This has a minor observable change - the return value is not normalized to 1/0/-1. This should be acceptable. We have done similar changes for other CompareTo methods in .NET Core.

File: src/System.Private.CoreLib/shared/System/Enum.cs [hunk 2]
```diff
@@ -906,6 +906,57 @@ public override string ToString()
...
+
+        public int CompareTo(object? target)
+        {
+            if (target == this)
+                return 0;
+
+            if (target == null)
+                return 1; // all values are greater than null
+
+            if (this.GetType() != target.GetType())
```
Comment (AaronRobinsonMSFT) [inline src/System.Private.CoreLib/shared/System/Enum.cs]: Although I prefer the `this.` prefix, I thought this was supposed to be avoided in product code according to (4) in the [coding guidelines](https://github.com/dotnet/corefx/blob/dc71f95667cabd5ad0f7a051c974e0d312b243e8/Documentation/coding-guidelines/coding-style.md#c-coding-style).

File: src/System.Private.CoreLib/src/System/Enum.CoreCLR.cs [hunk 3]
```diff
@@ -146,36 +146,5 @@ private static RuntimeType ValidateRuntimeType(Type enumType)
...
-
-        public int CompareTo(object? target)
-        {
-            const int retIncompatibleMethodTables = 2;  // indicates that the method tables did not match
-            const int retInvalidEnumType = 3; // indicates that the enum was of an unknown/unsupported underlying type
-
-            if (this == null)
-                throw new NullReferenceException();
-
-            int ret = InternalCompareTo(this, target);
```
Comment (pentp) [inline src/System.Private.CoreLib/src/System/Enum.CoreCLR.cs]: Could delete the `InternalCompareTo` method declaration from this file also.

Comment (jkotas) [reply src/System.Private.CoreLib/src/System/Enum.CoreCLR.cs]: https://github.com/dotnet/coreclr/pull/27837 Thanks!

**Pull request merged**